table/block_based_table_builder.cc

```diff
@@ -101,83 +101,106 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
-              compression_info,
-              GetCompressFormatForVersion(kBZip2Compression, format_version),
-              raw.data(), raw.size(), compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;  // fall back to no compression.
+      return BZip2_Compress(
+          compression_info,
+          GetCompressFormatForVersion(kBZip2Compression, format_version),
+          raw.data(), raw.size(), compressed_output);
     case kLZ4Compression:
-      if (LZ4_Compress(
-              compression_info,
-              GetCompressFormatForVersion(kLZ4Compression, format_version),
-              raw.data(), raw.size(), compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;  // fall back to no compression.
+      return LZ4_Compress(
+          compression_info,
+          GetCompressFormatForVersion(kLZ4Compression, format_version),
+          raw.data(), raw.size(), compressed_output);
     case kLZ4HCCompression:
-      if (LZ4HC_Compress(
-              compression_info,
-              GetCompressFormatForVersion(kLZ4HCCompression, format_version),
-              raw.data(), raw.size(), compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;     // fall back to no compression.
+      return LZ4HC_Compress(
+          compression_info,
+          GetCompressFormatForVersion(kLZ4HCCompression, format_version),
+          raw.data(), raw.size(), compressed_output);
     case kXpressCompression:
-      if (XPRESS_Compress(raw.data(), raw.size(),
-          compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;
+      return XPRESS_Compress(raw.data(), raw.size(), compressed_output);
     case kZSTD:
     case kZSTDNotFinalCompression:
-      if (ZSTD_Compress(compression_info, raw.data(), raw.size(),
-                        compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;     // fall back to no compression.
-    default: {}  // Do not recognize this compression type
+      return ZSTD_Compress(
+          compression_info, raw.data(), raw.size(), compressed_output);
+    default:
+      // Do not recognize this compression type
+      return false;
   }
+}
+
+}  // namespace
 
-  // Compression method is not supported, or not good compression ratio, so just
-  // fall back to uncompressed form.
+// format_version is the block format as defined in include/rocksdb/table.h
+Slice CompressBlock(
+    const Slice& raw,
+    const CompressionInfo& info,
+    CompressionType* type,
+    uint32_t format_version,
+    bool do_sample,
```

Comment:
The `do_sample` argument and `info.sample_for_compression` are redundant.

Reply:
do_sample is just passing in whether it is a data block or not. Sampling is rate. The caller could change the rate before passing in to 0 if they don't want to sample, and then change it back. But this seems cleaner.

Reply:
Should we rename it from `do_sample` to `is_data_block`? A sampling rate of `0` indicates sampling is disabled, doesn't it?

Reply:
Alex and I had talked about it. Overriding the sampling rate seemed undesirable. calling it do_sample allows the control to be in the caller. Calling it is_data_block would likely mean a common treatment for all callers. I'll take your recommendation for it.